cmd/snap-bootstrap/partition/encrypt.go

```diff
@@ -52,7 +58,43 @@ func (key EncryptionKey) Store(filename string) error {
 	// TODO:UC20: provision the TPM, generate and store the lockout authorization,
 	//            and seal the key. Currently we're just storing the unprocessed data.
 	if err := ioutil.WriteFile(filename, key[:], 0600); err != nil {
-		return fmt.Errorf("cannot store key file: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+type RecoveryKey [recoveryKeySize]byte
+
+func NewRecoveryKey() (RecoveryKey, error) {
+	var key RecoveryKey
+	// rand.Read() is protected against short reads
+	_, err := rand.Read(key[:])
+	// On return, n == len(b) if and only if err == nil
+	return key, err
+}
+
+// Store writes the recovery key in the location specified by filename.
+func (key RecoveryKey) Store(filename string) error {
+	if err := os.MkdirAll(filepath.Dir(filename), 0755); err != nil {
+		return err
+	}
+
+	groups := make([]string, 8)
+
+	// The recovery key is used as 8 groups of 5 base-10 digits, with each
+	// 5 digits being converted to a 2-byte number to make a 16-byte key.
+	r := bytes.NewReader(key[:])
+	for i := 0; i < 8; i++ {
+		var val uint16
+		if err := binary.Read(r, binary.LittleEndian, &val); err != nil {
+			return err
+		}
+		groups[i] = fmt.Sprintf("%05d", val)
+	}
+
```

Comment:
It's fine to just store the 16-byte key in binary format here. If we introduce any tooling later on for backing up the recovery key, the conversion to the numeric version can probably happen there if it needs to.

Reply:
It would also be useful to have something to regenerate the key in case it's compromised.

Reply:
Fixed this in PR #8103. I'll rebase on top of the new recovery key branch once it gets more reviews.